clients/da-vinci-client/src/main/java/com/linkedin/davinci/storage/StorageService.java

```diff
@@ -371,6 +376,43 @@ public synchronized AbstractStorageEngine openStore(
     return engine;
   }
 
+  public synchronized void checkWhetherStoragePartitionsShouldBeKeptOrNot(SafeHelixManager manager) {
+    if (manager == null) {
+      return;
+    }
+    for (AbstractStorageEngine storageEngine: getStorageEngineRepository().getAllLocalStorageEngines()) {
+      String storageEngineName = storageEngine.getStoreVersionName();
+      String storeName = Version.parseStoreFromKafkaTopicName(storageEngineName);
+      Set<Integer> storageEnginePartitionIds = storageEngine.getPartitionIds();
+      PropertyKey.Builder propertyKeyBuilder =
+          new PropertyKey.Builder(configLoader.getVeniceClusterConfig().getClusterName());
+      SafeHelixDataAccessor helixDataAccessor = manager.getHelixDataAccessor();
+      IdealState idealState = helixDataAccessor.getProperty(propertyKeyBuilder.idealStates(storeName));
+      String instanceHostName = manager.getInstanceName();
+
+      if (idealState != null) {
+        Map<String, Map<String, String>> mapFields = idealState.getRecord().getMapFields();
+        for (Map.Entry<String, Map<String, String>> entry: mapFields.entrySet()) {
+          boolean keepPartition = false;
+          String partitionDbName = entry.getKey();
+          int partitionId = RocksDBUtils.parsePartitionIdFromPartitionDbName(partitionDbName);
+          if (!storageEnginePartitionIds.contains(partitionId)) {
+            continue;
+          }
+          for (String hostName: entry.getValue().keySet()) {
+            if (hostName.equals(instanceHostName)) {
+              keepPartition = true;
+              break;
+            }
+          }
```

Comment:
Instead of looping over every key, can we instead check if the key `instanceHostName` exists in `entry`?